src/crypto/variant4_random_math.h

```diff
@@ -0,0 +1,405 @@
+			int alu_index = -1;
+			while (next_latency < TOTAL_LATENCY)
+			{
+				for (int i = op_ALUs[opcode] - 1; i >= 0; --i)
+				{
+					if (!alu_busy[next_latency][i])
+					{
+						// ADD is implemented as two 1-cycle instructions on a real CPU, so do an additional availability check
+						if ((opcode == ADD) && alu_busy[next_latency + 1][i])
+						{
+							continue;
+						}
+
+						// Rotation can only start when previous rotation is finished, so do an additional availability check
+						if (is_rotation[opcode] && (next_latency < rotate_count * op_latency[opcode]))
+						{
+							continue;
+						}
+
+						alu_index = i;
+						break;
+					}
+				}
+				if (alu_index >= 0)
+				{
+					break;
+				}
+				++next_latency;
+			}
+
+			// Don't generate instructions that leave some register unchanged for more than 7 cycles
+			if (next_latency > latency[a] + 7)
+			{
+				continue;
+			}
+
+			next_latency += op_latency[opcode];
+
+			if (next_latency <= TOTAL_LATENCY)
+			{
+				if (is_rotation[opcode])
+				{
+					++rotate_count;
+				}
+
+				// Mark ALU as busy only for the first cycle when it starts executing the instruction because ALUs are fully pipelined
+				alu_busy[next_latency - op_latency[opcode]][alu_index] = true;
+				latency[a] = next_latency;
+
+				// ASIC is supposed to have enough ALUs to run as many independent instructions per cycle as possible, so latency calculation for ASIC is simple
+				asic_latency[a] = ((asic_latency[a] > asic_latency[b]) ? asic_latency[a] : asic_latency[b]) + asic_op_latency[opcode];
+
+				rotated[a] = is_rotation[opcode];
+
+				inst_data[a] = code_size + (opcode << 8) + ((inst_data[b] & 255) << 16);
+
+				code[code_size].opcode = opcode;
+				code[code_size].dst_index = op.dst_index;
+				code[code_size].src_index = op.src_index;
+				code[code_size].C = 0;
+
+				if (opcode == ADD)
+				{
+					// ADD instruction is implemented as two 1-cycle instructions on a real CPU, so mark ALU as busy for the next cycle too
+					alu_busy[next_latency - op_latency[opcode] + 1][alu_index] = true;
+
+					// ADD instruction requires 4 more random bytes for 32-bit constant "C" in "a = a + b + C"
+					check_data(&data_index, sizeof(uint32_t), data, sizeof(data));
+					code[code_size].C = *((uint32_t*)&data[data_index]);
+					data_index += sizeof(uint32_t);
+				}
+
+				++code_size;
+				if (code_size >= NUM_INSTRUCTIONS)
+				{
+					break;
+				}
+			}
+			else
+			{
+				++num_retries;
+			}
+		}
+
+		// ASIC has more execution resources and can extract as much parallelism from the code as possible
+		// We need to add a few more MUL and ROR instructions to achieve minimal required latency for ASIC
+		// Get this latency for at least 1 of the 4 registers
+		const int prev_code_size = code_size;
+		while ((asic_latency[0] < TOTAL_LATENCY) && (asic_latency[1] < TOTAL_LATENCY) && (asic_latency[2] < TOTAL_LATENCY) && (asic_latency[3] < TOTAL_LATENCY))
```

Comment:
It's usually not more than asic_latency + 6-8 cycles.

Reply:
> It's usually not more than asic_latency + 6-8 cycles.

Is this suggesting that this loop typically adds 6-8 cycles of random code? Verifying that claim is a bit rough, but probably doable ...

> And we have to guarantee the minimal theoretical time required to run the generated random program, otherwise we'll have very easy code sequences sometimes.

If the minimum desired latency for an ASIC is too low, this generates more instructions with an 40+% skew towards multiplication. According to the logic within this PR, an ASIC has lower latency with such instructions. So this bottom portion is aggressively selecting instructions that are "slower" on a CPU, and is no longer taking the CPU execution ports into consideration. I'm not seeing a valid justification for this loop.

Reply:
Crud, hold on this loop generates code differently ...

Reply:
Actually, the skew towards MUL is even heavier in this second loop. Why? Take into consideration what I said two comments above this.

Reply:
ROR has latency of 2 cycles only on Intel in some extreme cases (several rotations in a row). ROR always has 1 cycle latency on Ryzen/Bulldozer, and it also has 1 cycle latency on Intel in this specific case (ror->mul->mul), so it's the same additional latency both for CPU and ASIC in this final sequence.

Reply:
> Actually, the skew towards MUL is even heavier in this second loop. Why?

This skew is by design. MUL is the preferred instruction of the whole random sequence, other instruction are just added to the mix to prevent optimizations of multiply sequences. Multiplication is basically the most optimized instruction on both CPU and GPU from latency point of view.

Reply:
> ROR has latency of 2 cycles only on Intel in some extreme cases (several rotations in a row). ROR always has 1 cycle latency on Ryzen/Bulldozer, and it also has 1 cycle latency on Intel in this specific case (ror->mul->mul), so it's the same additional latency both for CPU and ASIC in this final sequence.

I was asking about the general skew towards MUL, and even bigger skew in the bottom loop. Which you did discuss after that ...

> This skew is by design. MUL is the preferred instruction of the whole random sequence, other instructions are just added to the mix to prevent optimizations of multiply sequences. Multiplication is basically the most optimized instruction on both CPU and GPU from latency point of view.
>
> Hmm, maybe not on GPU, but CPU latency is the closest to possible multiplication latency on ASIC. GPUs can handle it thanks to their overall computing power.

If viewed pessimistically, there are 4 "root" dependency chains being shoved through a single MUL port on a CPU core. A specially designed chip could have multiple pipelines for handling this set of instructions. Ultimately the problem I have is that this design is fairly complex and design choices are not as "obvious" as the original cryptonote design.

Reply:
> there are 4 "root" dependency chains being shoved through a single MUL port on a CPU core. A specially designed chip could have multiple pipelines for handling this set of instructions.

Single MUL port on CPU can start new multiplication each cycle and this is taken into account by code generator. ASIC can of course do all independent instructions in parallel and this is also taken into account, which results in CPU latency being 6-8 cycles more than theoretical minimal latency for ASIC.

So ASIC will take at least 45 cycles to execute the program and CPU will take 51-53 cycles on average. But CPU usually takes around 80 cycles to run the rest of Cryptonight loop, so this difference doesn't play any role.

This design is complex only because I have to simulate how CPU works. If I just generated random code from a stream of data, difference between CPU and optimized ASIC would have been a lot more than 6-8 cycles in the end.

Reply:
There are multiple ports per CPU core for shifting, addition and division, which is why I've been probing on the bias towards multiply instructions. The choice is non-intuitive, and I'm still not seeing an _obvious_ advantage to that decision.

Reply:
I'll try to explain it from the different angle. Multiplication is the complex operation that requires O(logN) gate delay both for ASIC and CPU. They both would use similar implementation - https://en.wikipedia.org/wiki/Wallace_tree

The same implementation guarantees that ASIC won't be much faster than CPU here. Addition, subtraction have the same property - O(logN) gate delay: https://en.wikipedia.org/wiki/Brent%E2%80%93Kung_adder , but this delay is usually 3 times smaller than multiplication.

It would require 3 times more instructions to achieve the same execution delay, and CPU would run into instruction decoding stalls with such long programs and slow down unnecessarily. Moreover, 3 times more instruction where addition/subtraction are the most common enable much more opportunities for algebraic optimizations.

> There are multiple ports per CPU core for shifting, addition and division

We don't need (and don't want) high parallelism from generated code, because it's very easy for ASIC to outperform CPU in this area. Generated random code usually runs at 1.1-1.5 instructions per cycle on CPU, even with current bias towards multiplications, and this is enough. The main requirement for generated code is latency, not the total amount of instructions. The single port for multiplication can still start 1 multiplication every cycle, so it's possible to have 3 independent multiplications in flight at every moment. It's good enough when we only have 4 independent registers that can change.

Reply:
An example of algebraic optimization: `ADD(A, B) -> ADD(A, B)` can be optimized into `ADD(A, B*2)` where multiplication by 2 is simple shift. Two times lower latency here. Yes, compiler for CPU can also do this, but these optimizations make it impossible to guarantee latency for every generated program.

The similar sequence of multiplications `MUL(A, B) -> MUL(A, B)` can't be optimized for lower latency because `B^2=MUL(B,B) -> MUL(A, B^2)` has the same execution latency. Multiplication is more resistant to algebraic optimizations.

Reply:
One more problem with a bias towards additions and other simple instructions: longer programs that run at 2.5-3 instructions per cycle don't leave room for the rest of Cryptonight main loop which uses the same execution ports (it has only 1 multiplication), so it'll slow down 20-50% on CPU, but not on ASIC.